### ISSUE TITLE
componentWillMount used to register component into MVC, deprecated in react 16.3 / 17.x

### ISSUE THREAD

**Issue by jozefchutka** [reactjs/reactjs.org#724]
Hi,

in react 16.2 and older we used componentWillMount "event" to register component into the MVC framwork. That appeared as a perfect option to pass additional framework data as state into the component (doing it in componentDidMount caused extra render and flicker). Considering [componentWillMount deprication](https://reactjs.org/blog/2018/03/27/update-on-async-rendering.html) in 16.3+, what is the suggested event to register component into the framework?

cheers

**Comment by gaearon** [comment]
Please provide a full minimal example. It’s  not very clear what exactly you were doing.

Calling setState in componentDidMount applies the update synchronously and cannot cause flicker.

**Comment by jozefchutka** [comment]
Hi Dan,
what we are doing is hard to present with minimal example, but hopefully you can understand the important bits from the following:

```
class User extends React.Component {
  render() {
    return <div>{this.state.user}</div>;
  }
  componentWillMount(){
    this.props.myFramework.register(this);
  }
}

ReactDOM.render(
  <User myFramework="..."/>, document.getElementById('container')
);

class MyFramework {
  register(component){
    component.setState({user:"John"})
  }
}
```

Switching to the new version of React would enforce us to move component registration into componentDidMount and would cause [extra renderings](https://reactjs.org/docs/react-component.html#componentdidmount). Please consider its also about to re-render the whole tree of components when setState is called inside componentDidMount. 

I wonder if you have any recommendation how this can be optimized. Or there is really no option to avoid performance degradation (extra renderings) when using async version?

**Comment by gaearon** [comment]
Can you do this instead?

```js
class User extends React.Component {
  state =  this.props.myFramework.register()
  render() {
    return <div>{this.state.user}</div>;
  }
}

ReactDOM.render(
  <User myFramework="..."/>, document.getElementById('container')
);

class MyFramework {
  register(){
    return {user:"John"}
  }
}
```

I'd need more details about what the "framework" does to provide a better example but so far this is equivalent.

**Comment by gaearon** [comment]
Also, if you need to subscribe something (and update state in response to it) then [this should help](https://reactjs.org/blog/2018/03/27/update-on-async-rendering.html#adding-event-listeners-or-subscriptions) (don't miss the section on [create-subscription](https://github.com/facebook/react/tree/master/packages/create-subscription) package).

**Comment by jozefchutka** [comment]
Thanks for the suggestion Dan, but unfortunately this would lead to complete rewrite of our framework. Hopefully this can be avoided. Our framework is based on MVC pattern, more specifically on [robotlegs](http://www.robotlegs.org/) / [haxe-react-mmvc](https://github.com/elsassph/haxe-react-mmvc). The component can not really see the methods available on framework side, its the other way around, so framework class needs to control the component... I do not want to go into details as it might be way to much context for this issue.

Basically, from the framework side, we need setState to be available on component. What would be the earliest event in react 17 when setState is available? Is it really past the initial render?

cheers

**Comment by gaearon** [comment]
A small self-contained example that fully illustrates the problem would be necessary to continue. It's very hard to discuss this in abstract terms.

**Comment by jozefchutka** [comment]
Dan, here is a little more demonstration.  Language is haxe but should look familiar to you.

```
class User extends React.Component implements IUser {
  public var username(defaul, set):String;

  function set_username(value:String):String {
    setState({username:value});
  }

  function componentWillMount() {
    props.onViewAdded(this); // framework creates UserMediator and calls onRegister on it
  }

  function componentWillUnmount() {
    props.onViewRemoved(this); // framework calls onRemove on existing mediator and disposes it
  }
}

// Mediator is created by the framework for specific view components and has component instance available in property called view. Mediator can access other framework objects like models (data storage etc) and binds the data between view component and these objects.
class UserMediator extends Mediator<IUser> {
  @inject public var userModel:UserModel;

  function onRegister() {
    view.name = userModel.username;
    userModel.usernameChanged.add(onUsernameChanged);
  }

  function onRemove() {
    userModel.usernameChanged.remove(onUsernameChanged);
  }

  function onUsernameChanged() {
    view.name = userModel.username;
  }
}

interface IUser {
  var username(defaul, set):String;
}
```

In current implementation, concerns are separated and mvc side has no dependency over view component being used, just like view component has no idea what mvc framework is in use (just very thin layer onViewAdded/onViewRemoved). From mediator, we can set data without being interested in what lifecycle phase the view component is, so we do not need to distinguish if state is to be defined or setState called and yet the performance is optimal as there is only single initial render even for data passed from mediator.

Back to the original question, do you see any option how we can replace componentWillMount phase (used for mvc registration) by some other, where setState would be available already but at the same time would not cause additional initial rendering?

**Comment by gaearon** [comment]
As far as I can see you are trying to register a component instance in `componentWillMount` and subscribe it to some kind of external updates. This is exactly the pattern we encourage you to avoid because **it will lead to memory leaks with async rendering**. A single instance may be created and later “discarded” without any cleanup hook. 

Just because `componentWillMount` fired isn’t a guarantee that this instance will get mounted in the future (yes this is confusing, which is why we’re deprecating this lifecycle 🙂). The only appropriate place to make initial subscriptions is `componentDidMount`. 

So in any case you’d need to separate the “registration” phase (for which both constructor and componentWillMount are too early) and “initialization” (which should assign initial state and can happen in constructor). Does this help?

**Comment by jozefchutka** [comment]
Thanks Dan, that helps me understand an extra phase is needed for subscribing to external updates.

One more question regarding the topic. So if the life cycle is as following:

1. contstructor (define initial state)
2. async phase
3. render (consumes state)
4. setState is available

Considering some external updates happens right during this async phase (2), how to propagate these changes into the component? Should state be reinitialized or should setState() be called? Or there is no such option and new react is not capable of handling it?

**Comment by gaearon** [comment]
In your library you could have a helper that either assigns `this.state` when it doesn’t exist, or calls `setState` when it does.

But you’d still need to separate it in two stages.

**Comment by gaearon** [comment]
You’re right there may be an external update between the first render and mounting. I encourage you to look at `create-subscription` implementation (even if you don’t plan to use it) to see how to handle all edge cases. The recommendation here is to check if an external value changed in `componentDidMount`, and if it did, call `setState`. You don’t need to call it if the external value hasn’t changed.

https://github.com/facebook/react/blob/7a833dad95b3059ebfdfba44d3fa68e1301d8e6a/packages/create-subscription/src/createSubscription.js#L143

**Comment by jozefchutka** [comment]
Will do. Thanks!
I will close this issue now, I have collected all the info I wanted. Although, I think the new version does complicate things a bit more then it used to be, we will need to find a way around it on our side.

**Comment by gaearon** [comment]
We hope the benefits of async rendering will make it up for you 🙂

If you haven’t watched my talk yet, check it out: https://reactjs.org/blog/2018/03/01/sneak-peek-beyond-react-16.html
